### PR TITLE
Storage concurrently

### PR DESCRIPTION
**Pull request by liguozhong**
Storage concurrentl from #6716
issue https://github.com/prometheus/prometheus/issues/6714

### CODE REVIEW

File: storage/fanout.go [hunk 1]
```diff
@@ -240,27 +241,44 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
 func (q *mergeQuerier) SelectSorted(params *SelectParams, matchers ...*labels.Matcher) (SeriesSet, Warnings, error) {
 	seriesSets := make([]SeriesSet, 0, len(q.queriers))
 	var warnings Warnings
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(q.queriers))
```
Comment (brian-brazil) [inline storage/fanout.go]: You don't need the waitgrup any more, and the channel here doesn't even need to be buffered

Comment (liguozhong) [reply storage/fanout.go]: Thanks for your review , done.
Thank you for your help. I have less than 1 year experience in go language. My previous work experience is using Java

image

File: storage/fanout.go [hunk 2]
```diff
@@ -240,27 +240,40 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
...
-				return nil, nil, err
+				return nil, nil, qryResult.selectError
 			}
 		}
-		seriesSets = append(seriesSets, set)
 	}
 	return NewMergeSeriesSet(seriesSets, q), warnings, nil
 }
 
+type queryResult struct {
```
Comment (brian-brazil) [inline storage/fanout.go]: Put this inside the function, as you're only using it there.

File: storage/fanout.go [hunk 3]
```diff
@@ -240,27 +240,40 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
...
+			q.failedQueriers[qryResult.qr] = struct{}{}
 			// If the error source isn't the primary querier, return the error as a warning and continue.
-			if querier != q.primaryQuerier {
-				warnings = append(warnings, err)
-				continue
+			if qryResult.qr != q.primaryQuerier {
+				warnings = append(warnings, qryResult.selectError)
 			} else {
-				return nil, nil, err
+				return nil, nil, qryResult.selectError
```
Comment (brian-brazil) [inline storage/fanout.go]: If we return early here then the channel isn't emptied. We'll need a little more work here.

Comment (liguozhong) [reply storage/fanout.go]: done. A "priRrr" variable is added to ensure that the data in the channel can be consumed

image

File: storage/fanout.go (hunk 2 above)
Comment (liguozhong) [reply storage/fanout.go]: done. 

image

File: storage/fanout.go [hunk 4]
```diff
@@ -240,23 +240,46 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
...
+		selectError error
+	}
+	queryResultChan := make(chan *queryResult)
 	for _, querier := range q.queriers {
-		set, wrn, err := querier.SelectSorted(params, matchers...)
-		q.setQuerierMap[set] = querier
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
+		go func(qr Querier) {
+			if priErr != nil {
```
Comment (brian-brazil) [inline storage/fanout.go]: You've still a data race here, you can remove this if entirely.

File: storage/fanout.go [hunk 5]
```diff
@@ -240,23 +240,46 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
...
+				queryResultChan <- &queryResult{qr: qr, set: nil, wrn: nil, selectError: nil}
+				return
+			}
+			set, wrn, err := qr.SelectSorted(params, matchers...)
+			queryResultChan <- &queryResult{qr: qr, set: set, wrn: wrn, selectError: err}
+		}(querier)
+	}
+	for i := 0; i < len(q.queriers); i++ {
+		qryResult := <-queryResultChan
+		if priErr != nil {
```
Comment (brian-brazil) [inline storage/fanout.go]: We may aswell collect all the warnings when there's an error in the primary.

Comment (liguozhong) [reply storage/fanout.go]: done.
image

File: storage/fanout.go (hunk 4 above)
Comment (liguozhong) [reply storage/fanout.go]: Thank you for your review.
done.
image

File: storage/fanout.go [hunk 6]
```diff
@@ -240,23 +240,43 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
...
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
+		go func(qr Querier) {
+			set, wrn, err := qr.SelectSorted(params, matchers...)
+			queryResultChan <- &queryResult{qr: qr, set: set, wrn: wrn, selectError: err}
+		}(querier)
+	}
+	for i := 0; i < len(q.queriers); i++ {
+		qryResult := <-queryResultChan
+		if qryResult.wrn != nil {
```
Comment (brian-brazil) [inline storage/fanout.go]: You can simplify by removing this and the next `if`. The performance cost will be negligible.

Comment (liguozhong) [reply storage/fanout.go]: done.Thank you for your review.
image

**Comment by brian-brazil** [comment]
Thanks!

**Comment by liguozhong** [comment]
> Thanks!

I really very much appreciate your guidance and help.

**Comment by mknapphrt** [comment]
I tried building this to test it out and I get no results. Maybe I'm missing something but where are you actually appending the query results to `seriesSets `? AFAICT it makes an empty map at the beginning of the method and results that same empty map.

**Comment by brian-brazil** [comment]
Hmm, that looks to be the case and unittests didn't catch it either. Would you like to send a PR to fix it?

**Comment by mknapphrt** [comment]
I'll gladly PR the fix, all I did was
```
                qryResult := <-queryResultChan
                q.setQuerierMap[qryResult.set] = qryResult.qr
+              seriesSets = append(seriesSets, qryResult.set)
                if qryResult.wrn != nil {
                        warnings = append(warnings, qryResult.wrn...)
                }
```
but I don't really have time right now to work on the unit tests to go along with it.

**Comment by liguozhong** [comment]
I'll  work on the unit tests

**Pull request merged**